ffi/src/lib.rs

```diff
@@ -256,6 +256,42 @@ pub struct Opaque {
     __marker: core::marker::PhantomData<(*mut u8, core::marker::PhantomPinned)>,
 }
 
+/// Trait for `#[repr(transparent)]` structs to convert between [`Self`] and [`Self::Inner`]
+pub trait Transparent: Sized {
+    /// non-ZST field of transparent struct
+    type Inner;
+
+    /// Convert transparent struct into its non-ZST field
+    fn into_inner(outer: Self) -> Self::Inner;
+
+    /// Recover transparent struct from its non-ZST field
+    fn from_inner(inner: Self::Inner) -> Self;
+}
+
+/// Trait for `#[repr(transparent)]` structs to convert between slice of [`Self`] and slice of [`Self::Inner`]
+pub trait TransparentSlice: Sized {
```

Comment:
can't we replace this with:
```rs
impl<T: Transparent> iroha_ffi::Transparent for &[T] {
    type Inner = &[T::Inner];
}
```

or

```rs
impl<T: Transparent> iroha_ffi::Transparent for [T] {
    type Inner = [T::Inner];
}
```

depending on whether sizedness is required?

Reply:
because if `T` is transparent, doesn't that mean that `[T]` is also transparent?

Reply:
Answer for [T] i think yes, but for `Vec<T>` it's not correct because we doesn't have defined layout for the vector.
Thank you for the idea i will try to create such blanket impls.

Reply:
Ok, neither `Vec<T>` nor `&[T]` have defined layouts, but you can make blanket impls via pointer casts

Reply:
On 2nd thought I would implement `Transparent` only for `T`/`[T]` and add the following impls:

```rs
impl<T: ? Sized + Transparent<Inner = U>, U: ?Sized> IntoFfi for &T where &U: IntoFfi {
    type Target = <&U as IntoFfi>::Target:

    fn into_ffi(self) -> Self::Target {
        // do the transmutation
    }
}
```

but this might conflict with current blanket impl for `&T`

I can't verify this works right now, so it might be a blunt

Reply:
I'd recommend opening a ticket for it and adding the change later.

Reply:
> I think we should aim for this but this might conflict with current blanket impl for `&T`

Sadly there is conflict with `&T`.

Reply:
> I'd recommend opening a ticket for it and adding the change later.

Here it is #2668